### PR TITLE
Refactor MQ layer

### PR DESCRIPTION
**Pull request by ivan-bozhytksyi**
Treat it as v0.1.

* I created interfaces for Queue and Exchange. They are now stateful and not only strings. 
* Added pool to manage reconnection

### CODE REVIEW

File: network/mq/queue.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+			} else {
+				if err := msg.Ack(false); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+	}
+}
+
+func (q *queue) messageChannel() <-chan amqp.Delivery {
```
Comment (ivan-bozhytksyi) [inline network/mq/queue.go:98]: Use error instead of log.Fatal

File: network/mq/queue.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		false,
+		false,
+		false,
+		nil,
+	)
+	if err != nil {
+		log.Fatal("MQ issue" + err.Error() + " for queue: " + string(q.name))
+	}
+
+	err = q.amqpChan.Qos(
```
Comment (ivan-bozhytksyi) [inline network/mq/queue.go:112]: probably move it to setup

File: network/mq/pool.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+					log.Info("MQ Connection established")
+					break
+				}
+				log.Error("Failed to establish MQ connection: ", err)
+
+				if p.retriesNumber-i == 1 {
+					log.Fatal("MQ is not available now")
+				}
+
+				time.Sleep(p.timeout)
```
Comment (ivan-bozhytksyi) [inline network/mq/pool.go:104]: Increase time using Fibonacci sequence

File: network/mq/pool.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	conn, amqpChan, err := connect(url)
+	if err != nil {
+		log.Fatal("Failed to connect to AMQP: ", err)
+	}
+
+	pool := &Pool{
+		connURL:  url,
+		amqpChan: amqpChan,
+		conn:     conn,
+
```
Comment (ivan-bozhytksyi) [inline network/mq/pool.go:39]: add default retries number

File: network/mq/mq.go [hunk 5]
```diff
@@ -1,38 +1,27 @@
...
 
 	log "github.com/sirupsen/logrus"
-
 	"github.com/streadway/amqp"
 )
 
 var (
-	amqpChan *amqp.Channel
-	conn     *amqp.Connection
+	Conn     *amqp.Connection
```
Comment (ivan-bozhytksyi) [inline network/mq/mq.go:13]: do not expose.

File: network/mq/options.go [hunk 6]
```diff
@@ -19,3 +19,9 @@ func InitDefaultConsumerOptions(workers int) ConsumerOptions {
 		RetryDelay:    time.Second * 1,
 	}
 }
+
+func PoolOptionRetriesNumber(number int) PoolOption {
```
Comment (vikmeup) [inline network/mq/options.go:23]: let's implement constant reconnect, 1, 3, 10, 15, 30 seconds periods?

File: network/mq/pool.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		timeout: time.Second * 10, // default timeout value
```
Comment (vikmeup) [inline network/mq/pool.go:40]: can we add timeout as an option?

File: network/mq/pool.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+func (p *Pool) Close() error {
+	err := p.amqpChan.Close()
+	if err != nil {
+		log.Print(err)
+	}
+
+	return p.conn.Close()
+}
+
+func (p *Pool) Start(ctx context.Context) {
```
Comment (vikmeup) [inline network/mq/pool.go:77]: maybe pass queues on start?

**Review by vikmeup — APPROVED** [review]

**Review by vikmeup — APPROVED** [review]

File: network/mq/pool.go [hunk 9]
```diff
@@ -57,12 +65,11 @@ func (p *pool) Consume(ctx context.Context) error {
 		if p.mq.conn.IsClosed() {
 			log.Warn("MQ connection lost")
 
-			if p.retriesNumber == 0 {
-				log.Fatal("MQ connection closed")
-			}
+			for i := 0; i < len(retriesDelay); i++ {
```
Comment (vikmeup) [inline network/mq/pool.go:68]: for simplicity maybe implement each 15 or 30 sec retry? should cover most use cases

File: network/mq/mq.go [hunk 10]
```diff
@@ -43,141 +32,48 @@ func (c ConsumerDefaultCallback) Callback(msg amqp.Delivery) error {
 	return c.Delivery(msg)
 }
 
-func Close() error {
-	err := amqpChan.Close()
+func Connect(url string) {
```
[the remark this answers is not in the document]
Comment (ivan-bozhytksyi) [reply network/mq/mq.go]: The client should be able to close it explicitly. 
Tried to manage closing through context - adds a lot of complexity and is non obvious for client apps.

**Comment by ivan-bozhytksyi** [comment]
@vikmeup added closing via context, works well

**Pull request merged**